chichiwang/exercise-js
# Add Sass modules

## Description
Linked to Issue: #20 
[Sass](https://sass-lang.com/) is instrumented into the project using [CSS modules](https://github.com/css-modules/css-modules). This will allow modular styling via CSS within the boilerplate.

This PR introduces a solution using [style-loader](https://webpack.js.org/loaders/style-loader/), which is [not recommended for production builds](https://webpack.js.org/loaders/style-loader/#security-warning). This will not be addressed in Issue #20, it is better suited to be handled in Issue #45.

## Changes
* Add devDependencies to `package.json`
  * `css-loader` - process CSS after `sass-loader` transpiles SASS to CSS
  * `identity-obj-proxy` - transforms CSS module imports into object properties for Jest to handle CSS modules
  * `sass` - necessary for `sass-loader` as a peer dependency
  * `sass-loader` - process SCSS modules into CSS
  * `style-loader` - writes style rules into the DOM, for use in development modes
  * `typescript-plugin-css-modules` - Allows TS to understand CSS modules, and allows auto type definitions for CSS modules to be generated
* `.eslintrc.js` configured to update rule for `filenames/match-regex`
  * Allow for `.` to be present in a filename for files such as `global.d.ts` or `*.module.scss`
* `app/App` imports global styles module `app/global.module.scss`
* Add file `app/global.module.scss` to house global styles
  * One rule created: anchor tags
* Add file `app/global.d.ts` to house global type definitions
  * Declare modules `*.module.scss` to be auto-defined by plugin `typescript-plugin-css-modules`
* Add file `app/var.module.scss` to house app-level global variables for use
  * Define one [map](https://sass-lang.com/documentation/values/maps/) `$colors` to house hex codes for colors in the app
* Add configuration to `jest.config.js`
  * Entry added to `moduleNameMapper` to css module files to be handled by `identity-obj-proxy`
* `pages/Home` updated to import styles from `pages/Home/styles.module.scss`
  * Apply a class `.heading` to `<h1>` heading on the page
* `pages/Home/styles.module.scss` created to define styles for `pages/Home`
* `tsconfig.json` updated to add a plugin `typescript-plugin-css-modules`
* `webpack-config/base.js` updated to add configuration to process `.scss`/`.sass` files

## Steps to QA
* Pull down and switch to this branch
* Run `npm run watch` to ensure assets are built correctly and watch server starts
  * Navigate to it in the browser and ensure styles are reflected
  * Inspect heading or anchor tag on home page to ensure styles are properly source-mapped to original code
* Run `npm run build:devRun` to ensure assets are built correctly and watch server starts
  * Navigate to it in the browser and ensure styles are reflected
  * Inspect heading or anchor tag on home page to ensure styles are properly source-mapped to original code
